src/lib.rs

```diff
@@ -1,289 +1,69 @@
     }
-
-    pub struct Tree<'a, T> {
-        v: T,
-        children: Stream<'a, Tree<'a, T>>
-    }
 }
 
 #[cfg(test)]
 mod tests {
-    use lazy::{Thunk, ConsCell, Stream};
+    use lazy::{Thunk, Tree};
     use std::time::SystemTime;
-    //use test::Bencher;
 
     #[test]
     fn thunks_defer_application_until_forced() {
-        let t = Thunk::new(|| 5);
-        let v = t.force();
-        assert_eq!(v, 5);
+        let mut t = Thunk::new(|| SystemTime::now());
+        let v = t.force().value;
+        assert!(v != Some(SystemTime::now()));
     }
 
     #[test]
     fn thunks_memoize_values() {
-        let t = Thunk::new(|| SystemTime::now());
-        let p = t.force();
-        let q = t.force();
+        let mut t = Thunk::new(|| SystemTime::now());
+        let p = t.force().value;
+        let q = t.force().value;
         assert_eq!(p, q);
     }
 
     #[test]
-    fn streams_are_lazy_and_possibly_infinite() {
-        fn ints_from<'a>(n: usize) -> Stream<'a, usize> {
-            Stream::new(move || ConsCell::new(n, ints_from(n+1)))
-        }
-
-        let mut strm = ints_from(5);
-        let mut i = 5;
-        loop {
-            if i > 10 {
-                break;
-            }
-
-            assert_eq!(strm.get(), Some(i));
-            i += 1;
-            strm = strm.tail();
-        }
-    }
-
-    #[test]
-    fn streams_are_lazy_and_possibly_finite() {
-        fn ints_from_to<'a>(n: usize, m: usize) -> Stream<'a, usize> {
-            if n > m {
-                return Stream::empty();
-            }
-            Stream::new(move || ConsCell::new(n, ints_from_to(n+1, m)))
-        }
-
-        let mut strm = ints_from_to(5, 7);
-        let mut i = 5;
-        loop {
-            if i > 7 {
-                assert_eq!(strm.get(), None);
-                break;
-            }
-
-            assert_eq!(strm.get(), Some(i));
-            i += 1;
-            strm = strm.tail();
-        }
-    }
-
-    #[test]
-    fn infinite_streams_impl_iterators() {
-        fn ints_from<'a>(n: usize) -> Stream<'a, usize> {
-            Stream::new(move || ConsCell::new(n, ints_from(n+1)))
-        }
-
-        let strm = ints_from(5);
-        let mut i = 5;
-        for x in strm {
-            if i > 100 {
-                break;
-            }
-
-            assert_eq!(x, i);
-            i += 1;
-        }
-    }
-
-    #[test]
-    fn finite_streams_impl_iterators() {
-        fn ints_from_to<'a>(n: usize, m: usize) -> Stream<'a, usize> {
-            if n > m {
-                return Stream::empty();
-            }
-            Stream::new(move || ConsCell::new(n, ints_from_to(n+1, m)))
-        }
-
-        let strm = ints_from_to(5, 7);
-        let mut i = 5;
-        for x in strm {
-            assert_eq!(x, i);
-            i += 1;
-        }
+    fn rose_trees_hold_lazy_values() {
+        let mut tree = Tree::singleton(42);
+        let p = tree.thunk.force().value;
+        let q = tree.thunk.force().value;
```

Comment:
just from a design point of view i think this would be nicer as `tree.thunk.force()` is that possible in the type system? or even better `tree.value()` which forces the thunk.

Reply:
Great point, I think I'm going to try to rehash this a bit to make that happen before this gets merged